nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/hazelcast/encryption/SymmetricEncryptor.java

```diff
@@ -0,0 +1,222 @@
+            throw new HazelcastException("Error reading encrypted key", ioe);
+        }
+
+        if (encryptedBytes == null) {
+            throw new HazelcastException("Encrypted key appears to be null");
+        }
+
+        try {
+            Cipher cipher = Cipher.getInstance(AES_ALGORITHM);
+
+            // Strip off the salt and IV
+            ByteBuffer buffer = ByteBuffer.wrap(Base64.getDecoder().decode(encryptedBytes));
+            byte[] saltBytes = new byte[20];
+            buffer.get(saltBytes, 0, saltBytes.length);
+            byte[] ivBytes = new byte[cipher.getBlockSize()];
+            buffer.get(ivBytes, 0, ivBytes.length);
+            byte[] encryptedTextBytes = new byte[buffer.capacity() - saltBytes.length - ivBytes.length];
+            buffer.get(encryptedTextBytes);
+
+            // Deriving the key
+            SecretKeyFactory factory = SecretKeyFactory.getInstance(PBKDF_ALGORITHM);
+            PBEKeySpec spec = new PBEKeySpec(masterPassword, saltBytes, ITERATION_COUNT, KEYSIZE);
+            SecretKey secretKeySpec = new SecretKeySpec(factory.generateSecret(spec).getEncoded(), AES);
+            cipher.init(Cipher.DECRYPT_MODE, secretKeySpec, new IvParameterSpec(ivBytes));
+            return new SecretKeySpec(cipher.doFinal(encryptedTextBytes), "AES");
+        } catch (BadPaddingException bpe) {
+            throw new HazelcastException("BadPaddingException caught decrypting data grid key" +
+                    "- likely caused by an incorrect or changed master password", bpe);
+        } catch (IllegalBlockSizeException | NoSuchAlgorithmException | InvalidAlgorithmParameterException
+                | InvalidKeyException | InvalidKeySpecException | NoSuchPaddingException exception) {
+            throw new HazelcastException(exception);
+        }
+    }
+
+    public static byte[] objectToByteArray(Object object) {
+        byte[] bytes = null;
+        try (ByteArrayOutputStream bos = new ByteArrayOutputStream(); ObjectOutputStream out = new ObjectOutputStream(bos)) {
+            out.writeObject(object);
+            bytes = bos.toByteArray();
+        } catch (IOException ioe) {
+            // See "if (bytes == null)"
+        }
+        if (bytes == null) {
+            throw new HazelcastException("Error converting Object to Byte Array");
+        }
+        return bytes;
+    }
+
+    public static Object byteArrayToObject(byte[] bytes) {
+        Object object;
+        try (ByteArrayInputStream bis = new ByteArrayInputStream(bytes)) {
+            object = IOUtil.newObjectInputStream(Thread.currentThread().getContextClassLoader(), null, bis).readObject();
+        } catch (IOException | ClassNotFoundException exception) {
+            throw new HazelcastException("Error converting Byte Array to Object", exception);
+        }
+        if (object == null) {
+            throw new HazelcastException("Object appears to be null, something probably went wrong converting Byte Array to Object");
```

Comment:
What about encrypting `null` references?
If I read code in store correctly nulls are not wrapped and stored "plain".

Reply:
Good spot!